### ISSUE TITLE
support for ForwardRef (forward references)

### ISSUE THREAD

**Issue by buhman** [samuelcolvin/pydantic#234]
I expected this would do something sane:

```python
from pydantic import BaseModel
from typing import Optional, ForwardRef

Foo = ForwardRef('Foo')
class Foo(BaseModel):
    foo: Optional[Foo]
```

Instead it does not:

```pytb
Traceback (most recent call last):
  File "./lib/python3.7/site-packages/pydantic/validators.py", line 262, in find_validators
    if issubclass(type_, val_type):
TypeError: issubclass() arg 1 must be a class

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./test.py", line 6, in <module>
    class Foo(BaseModel):
  File "./lib/python3.7/site-packages/pydantic/main.py", line 120, in __new__
    config=config,
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 105, in infer
    schema=schema,
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 85, in __init__
    self.prepare()
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 133, in prepare
    self._populate_sub_fields()
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 177, in _populate_sub_fields
    ) for t in types_]
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 177, in <listcomp>
    ) for t in types_]
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 85, in __init__
    self.prepare()
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 134, in prepare
    self._populate_validators()
  File "./lib/python3.7/site-packages/pydantic/fields.py", line 217, in _populate_validators
    self.model_config.arbitrary_types_allowed)),
  File "./lib/python3.7/site-packages/pydantic/validators.py", line 265, in find_validators
    raise RuntimeError(f'error checking inheritance of {type_!r} (type: {display_as_type(type_)})') from e
RuntimeError: error checking inheritance of ForwardRef('Foo') (type: ForwardRef('Foo'))
```

It looks like ForwardRef isn't resolved (and can't be, in this case, because it isn't defined yet).

**Comment by samuelcolvin** [comment]
Probably possible to implement. 

Will look in a week or two.

**Comment by demosdemon** [comment]
I started looking into this an also discovered that pydantic does not handle py37's annotations future import either.

```python
from __future__ import annotations

from pydantic import BaseModel

class Foo(BaseModel):
    x: str = ''
```

The `annotations` future import causes the python engine to leave annotations as `str` values instead of interpreting them. It is up to the user of the annotations to call [`typing.get_type_hints`](https://github.com/python/cpython/blob/dcf14d1a9875143747cd87c66fae6e693b21c6de/Lib/typing.py#L926) or cast them to `FutureRef` objects and invoke the [`_evaluate`](https://github.com/python/cpython/blob/dcf14d1a9875143747cd87c66fae6e693b21c6de/Lib/typing.py#L455) method. However, this is not a public API as opposed to `get_type_hints`. 

Unfortunately, `get_type_hints` requires that the object has already been created. Pydantic [currently accesses](https://github.com/samuelcolvin/pydantic/blob/d3cec7be5693276a0449c466cfd64495a42563d5/pydantic/main.py#L135) the `__annotations__` structure during the metaclass `__new__` method, which is _just_ before the object is created and inserted into the module namespace; this makes any self-references impossible resolve at the time:

```python
from typing import Optional
from pydantic import BaseModel

class Foo(BaseModel):
    x: Optional['Foo'] = None
```

The name `Foo` in the example above does not exist until our metaclass `__new__` method returns.

To support the etherial Python 4.0 as well as current `FutureRef` instances, resolving and validating type hints must occur after instantiating the class. Perhaps this should be deferred to the [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) method?

**Comment by samuelcolvin** [comment]
@demosdemon thanks for the detailed explanation, it helped me a lot.

This should now be fixed in #348, however since I'm not currently a big user of `ForwardRef` and am not entirely sure of all the use cases, I'd appreciate it if anyone wanting this feature, could try this code.

Any feedback on #348 please.

@buhman @PaulKuiper @demosdemon @koliyo @dmfigol
